From 486732f0374928fc77485111ec51148ef144e8a0 Mon Sep 17 00:00:00 2001
From: Abinaya Kalaiselvan <akalaise@codeaurora.org>
Date: Fri, 6 Nov 2020 13:51:05 +0530
Subject: [PATCH] ath11k: support for WMI service bit beyond 256

Introduce wmi_ext2_service_bitmap for handling WMI service
bit beyond 256.

Signed-off-by: Abinaya Kalaiselvan <akalaise@codeaurora.org>
---
 drivers/net/wireless/ath/ath11k/wmi.c | 23 +++++++++++++++++++++++
 drivers/net/wireless/ath/ath11k/wmi.h |  4 +++-
 2 files changed, 26 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath11k/wmi.c b/drivers/net/wireless/ath/ath11k/wmi.c
index e8039f8..cf3e159 100644
--- a/drivers/net/wireless/ath/ath11k/wmi.c
+++ b/drivers/net/wireless/ath/ath11k/wmi.c
@@ -7222,6 +7222,7 @@ static void ath11k_service_available_event(struct ath11k_base *ab, struct sk_buf
 {
 	const void **tb;
 	const struct wmi_service_available_event *ev;
+	u32 *wmi_ext2_service_bitmap;
 	int ret;
 	int i, j;
 
@@ -7258,6 +7259,28 @@ static void ath11k_service_available_event(struct ath11k_base *ab, struct sk_buf
 		   ev->wmi_service_segment_bitmap[0], ev->wmi_service_segment_bitmap[1],
 		   ev->wmi_service_segment_bitmap[2], ev->wmi_service_segment_bitmap[3]);
 
+	wmi_ext2_service_bitmap = (u32 *) tb[WMI_TAG_ARRAY_UINT32];
+
+	if (!wmi_ext2_service_bitmap) {
+		ath11k_warn(ab, "failed to fetch svc available wmi_ext2_service_bitmap");
+		kfree(tb);
+		return;
+	}
+
+	for (i=0, j = WMI_MAX_EXT_SERVICE; j < WMI_MAX_EXT2_SERVICE;
+	     i++) {
+		do {
+			if (wmi_ext2_service_bitmap[i] &
+			    BIT(j % WMI_AVAIL_SERVICE_BITS_IN_SIZE32))
+				set_bit(j, ab->wmi_ab.svc_map);
+		} while (++j % WMI_AVAIL_SERVICE_BITS_IN_SIZE32);
+        }
+
+	ath11k_dbg(ab, ATH11K_DBG_WMI,
+		   "wmi_ext2_service_bitmap 0:0x%x, 1:0x%x, 2:0x%x, 3:0x%x",
+		   wmi_ext2_service_bitmap[0], wmi_ext2_service_bitmap[1],
+		   wmi_ext2_service_bitmap[2], wmi_ext2_service_bitmap[3]);
+
 	kfree(tb);
 }
 
diff --git a/drivers/net/wireless/ath/ath11k/wmi.h b/drivers/net/wireless/ath/ath11k/wmi.h
index a37ce69..6c25d37 100644
--- a/drivers/net/wireless/ath/ath11k/wmi.h
+++ b/drivers/net/wireless/ath/ath11k/wmi.h
@@ -2140,6 +2140,8 @@ enum wmi_tlv_service {
 	WMI_TLV_SERVICE_CONFIGURE_ROAM_TRIGGER_PARAM_SUPPORT = 254,
 	WMI_TLV_SERVICE_CFR_TA_RA_AS_FP_SUPPORT = 255,
 
+	WMI_MAX_EXT_SERVICE = 256,
+
 	WMI_TLV_SERVICE_CFR_CAPTURE_COUNT_SUPPORT = 256,
 	WMI_TLV_SERVICE_OCV_SUPPORT = 257,
 	WMI_TLV_SERVICE_LL_STATS_PER_CHAN_RX_TX_TIME_SUPPORT = 258,
@@ -2149,7 +2151,7 @@ enum wmi_tlv_service {
 	WMI_TLV_SERVICE_FSE_CMEM_ALLOC_SUPPORT = 262,
 	WMI_TLV_SERVICE_PASSIVE_SCAN_START_TIME_ENHANCE = 263,
 
-	WMI_MAX_EXT_SERVICE
+	WMI_MAX_EXT2_SERVICE
 };
 
 enum wmi_unit_test_cmdid {
-- 
2.7.4

